### ISSUE TITLE
Golang Support

### ISSUE THREAD

**Issue by joegrasse** [oracle/graal#381]
Will Go be supported in the future?

**Comment by thomaswue** [comment]
It could be possible via a Go => LLVM => GraalVM bridge. There is this project https://go.googlesource.com/gollvm/ and this project https://github.com/go-llvm/llgo. The former looks more up-to-date. We will investigate what it takes.

**Comment by eginez** [comment]
For what is worth I tried running lli with https://github.com/go-llvm/llgo and found that lli was failing to load some of the native libraries. I created a docker image that can be used for testing `docker pull eginez/graalgo`, fair warning is a big image(~17Gb)

**Comment by lukasstadler** [comment]
Yes, the relationship between the gollvm-created bitcode and the Go runtime library is somewhat special - FWIW, even llvm's own "lli" seems to be unable to handle it.
We're looking into various ways to make that work, we may need to introduce a special hook for Go code.

**Comment by CAFxX** [comment]
> It could be possible via a Go => LLVM => GraalVM bridge. 

Wouldn't going through truffle be better than LLVM in this case? Compiled Go programs contain a pretty big/multithreaded runtime (including a user-mode scheduler tightly integrated with the garbage collector and the built-in I/O event loop) that would need to run on top of the graal runtime. Even if suboptimal going the LLVM way may work at least in simple cases (footgun number one is likely going to be the presence of two GCs) but it would be pretty unusable for anything non-trivial involving cross-language interop (that would also depend on cgo, since there's no other supported way of doing interop with the standard go runtime). With a "trufflego" approach these issues wouldn't apply - although it would require signifcant work to get them right.

(I'm gleefully glossing over the fact that going with truffle would require reimplementing a significant part of the runtime, as well as a Go interpreter -- and that therefore the amount of work is not even remotely comparable)

**Comment by eginez** [comment]
golang is about release a compiler that can target `wasm`, I wonder if we could use that target and load it through the graalvm

**Comment by ScalaWilliam** [comment]
I'm interested in this.

**Comment by thomaswue** [comment]
We did get more requests for this feature recently. If anybody has some concrete example scenarios in mind for running Go on GraalVM; e.g., what would be the main expected benefits (interoperability? performance? tooling?), please share.

**Comment by chengenzhao** [comment]
> We did get more requests for this feature recently. If anybody has some concrete example scenarios in mind for running Go on GraalVM; e.g., what would be the main expected benefits (interoperability? performance? tooling?), please share.

We do have an example
currently we are using Vert.X toolkit and we hope we could use Goroutine in the verticle just like Kotlin Coroutine, which will make async code relatively easy to write.
For Kotlin Coroutine, we have to launch the coroutine scope, but in Golang, we dont have to do that, we could go keywork to simplify the async code, Java has similar feature which is Project Loom but it takes time to complete, so if we could write the code in Golang then the vertx-lang-golang/go4x would be a perfect match in this scenario.

**Comment by chengenzhao** [comment]
Vert.X has a language support based on Graal, which is es4x project
https://reactiverse.io/es4x/
and we could create similar projects to use Golang e.g. go4x
we do have many people experienced in vertx-codegen tool
so if Graal could run Go language, we could generate vertx-lang-go/go4x to wrap vertx apis
and provide these apis to Go users, for now, about green thread/fiber/coroutine etc. 
Golang is the easiest way to write, we could benefits from this feature
A similar scenario is Kotlin Coroutine support which makes Kotlin code easy to write in Vert.x
e.g.
https://vertx.io/docs/vertx-lang-kotlin-coroutines/kotlin/
so we could generate a similar feature for Vert.X

**Comment by marcellanz** [comment]
> We did get more requests for this feature recently. If anybody has some concrete example scenarios in mind for running Go on GraalVM; e.g., what would be the main expected benefits (interoperability? performance? tooling?), please share.

@thomaswue it would lower the bar entering go applications in environments where the JVM is traditionally established, both by expertise, tooling and its operational foundation. think about companies that restrict their decision to use a language if it runs on the JVM.

**Comment by eginez** [comment]
I think a killer use case is just to leverage libraries written in Java from Go and viceversa. The java ecosystem is very rich in the open source as well as within corporations. Allowing people to run software in go(or java) without having to re-implement well test libraries would be key feature for graalvm. This is specially true given go is gaining grounds as a backend language.

**Comment by joegrasse** [comment]
My use case, and the reason I opened the original issue, was so that I can use Go to write stored programs in MySQL, instead of [javascript](https://www.percona.com/live/18/sessions/running-javascript-stored-programs-inside-mysql-server).

**Comment by eginez** [comment]
here is some experimental solution for this issue: https://gist.github.com/eginez/d69e3e1c4045eb692a274bd2bd5d31d6

**Comment by thomaswue** [comment]
How this tre project compare to https://go.googlesource.com/gollvm/?

**Comment by eginez** [comment]
agreed the above post is  demonstrative only

**Comment by spangaer** [comment]
Being able to parse a terraform script in to JVM using it's original parser implementation would be a nice use case.
;)

**Comment by eginez** [comment]
The support for WASM in GraalVM opens yet another door to run go in the JVM. This time we would via go's wasm backend.

**Comment by smasher164** [comment]
How does support for Project Loom affect the viability of a truffle implementation for the Go runtime?

**Comment by eginez** [comment]
It would surely make the implementation of go-routines much easier

**Comment by CAFxX** [comment]
> If anybody has some concrete example scenarios in mind for running Go on GraalVM; e.g., what would be the main expected benefits (interoperability? performance? tooling?), please share.

One scenario is that of compute-heavy applications; the go compiler is heavily focused on compilation speed, and for this reason it is extremely conservative about which optimizations are performed (inlining is done extremely sparingly, very little devirtualization is supported, no LICM, no autovectorization, very basic native arch detection, and especially no support for PGO). As a result, compute-heavy operations (e.g. [gzip encoding/deecoding](https://github.com/golang/go/issues/23154)) when implemented in go and exposed with idiomatic interfaces end up being slower.

GCC and llvmgo could help address some of these points, but IIRC they are missing higher-level go-specific knowledge about the memory allocator, GC and goroutine stacks, so they end up allocating more on the heap compared to the go compiler. 

At least on paper, a truffle-based implementation of go should fare significantly better here, especially for server-like workloads where there would be plenty of time to specialize the compiled code.

**Comment by zonghaishang** [comment]
> 最近，我们确实收到了更多关于此功能的请求。如果有人想到了在GraalVM上运行Go的一些具体示例场景；例如，请分享主要的预期收益（互操作性？性能？工具？）。

At present, the wasm module is supported on the go project mosn, and we hope that the wasm compiled by java can also be executed (outside the js environment).

We have implemented protocol extensions in the rpc field, so we also hope that java developers can provide extension plugins.

We noticed that wasm currently supports multiple language implementations：https://github.com/proxy-wasm/spec

- AssemblyScript SDK
- C++ SDK
- Go (TinyGo) SDK
- Rust SDK

In the java compiler, no examples of use in a non-web environment have been found so far.

**Comment by dcorbe** [comment]
LLVM supports GCC's IR.  Has anyone investigated the possibility of using gccgo as a bridge?

**Comment by ShalokShalom** [comment]
The wasm backend of Go, does seem to be production ready now: 

https://github.com/golang/go/wiki/WebAssembly

**Comment by ShalokShalom** [comment]
And I found also a binding to LLVM, although it speaks about a 'system-installed' LLVM, and is at least hosted by the TinyGo project, so it may provide just bindings for this subset

https://github.com/tinygo-org/go-llvm

**Comment by dgutson** [comment]
> We did get more requests for this feature recently. If anybody has some concrete example scenarios in mind for running Go on GraalVM; e.g., what would be the main expected benefits (interoperability? performance? tooling?), please share.

We have an SDK, with different codebases in different languages. Currently, we have to replicate the SDK into each language we support (support madness); I would like to have one SDK in one language, and use GraalVM to interop with the user code. Bindings are also a mess.

This being said, also supporting haskell would be so useful.

**Comment by ShalokShalom** [comment]
> > We did get more requests for this feature recently. If anybody has some concrete example scenarios in mind for running Go on GraalVM; e.g., what would be the main expected benefits (interoperability? performance? tooling?), please share.
> 
> We have an SDK, with different codebases in different languages. Currently, we have to replicate the SDK into each language we support (support madness); I would like to have one SDK in one language, and use GraalVM to interop with the user code. Bindings are also a mess.
> 
> This being said, also supporting haskell would be so useful.

I did not try it for myself, and remember people saying it does run on Graal Sulong. 

The Haskell compiler has a native LLVM backend these days.

**Comment by Zireael07** [comment]
You asked for concrete examples: I have some Go scripts (image processing) - in Go because I can't write C and Go is almost as performant AND it compiles fast. But Go doesn't have good visualization libraries, while e.g. Python and Javascript do (matplotlib and d3)

**Comment by ShalokShalom** [comment]
Did you see the wasm and llvm compiler? It seems like an opportunity to run Go on Graal.
